compressor: Update readme (#18395)
microsoft/FluidFramework: packages/runtime/container-runtime/src/id-compressor/README.md

# **Distributed ID Allocator**

A library which generates small number representations of arbitrary non-colliding Version 4 UUIDs (stable IDs) across multiple sessions in a network. This scheme enables a distributed application to utilize the global uniqueness guarantees of UUIDs while maintaining the performance advantages of small integers.

## **Overview**

> The distributed ID allocation scheme allows clients to use, store, and reference small numbers that map to UUIDs. The primary benefits are improved memory efficiency and reduced storage size, as well as improved runtime performance when used as keys in collections or other data structures in place of UUIDs.

Reviewer: ```suggestion
The distributed ID allocation scheme allows clients to use, store, and reference small numbers that map to UUIDs rather than the UUIDs themselves. The primary benefits are improved memory efficiency and reduced storage size, as well as improved runtime performance when used as keys in collections.
```
To slightly clarify the overall purpose in the first sentence rather than waiting until the second.

The scheme requires a total order for ID allocation, so the allocator is designed to work within a service that provides centralized total order broadcast.

## **Sessions**

A session can be thought of as a unique identifier for a single allocator. Even if an allocator is taken offline, serialized to disk, and rehydrated back, it's considered the same session if it has the same session ID.

## **Document**

The set of IDs generated by all Sessions contributing to the same distributed collection (e.g. an individual collaborative document). Uniqueness of allocated IDs is limited to the document context (unless in their UUID form), and thus IDs in their compressed form are not portable between documents.

## **Generation API**

> New IDs are generated via synchronous API on the allocator. The API returns a small number representation of the UUID with the following guarantees:

Reviewer: ```suggestion
New IDs are generated via a synchronous API on the allocator. The API returns a small number representation of the UUID with the following guarantees:
```
@alex-pardes Asynchronous collaboration for a synchronous collaboration environment 12/11/17

- The ID can be synchronously converted into a full UUID at any time. This is required if global uniqueness (across documents) is needed.

> - Prior to the service being notified of the ID's creation, it is unique within the scope of the session (e.g. it can collide with IDs allocated by a different allocator in the same network).

Reviewer: ```suggestion
- Prior to the service being notified of the ID's creation, it is unique within the scope of the session (i.e. the new ID will not collide with other IDs from the same allocator, but it can collide with IDs from a different allocator in the same network).
```
Statement in parentheses confused me because it's clarifying the preceding point, but doing it by counterexample without indicating so

> - After the service has been notified of the ID's creation and has determined a total ordering between all clients creating IDs, the ID will have a _final_ form that is a small number representation that is unique across the [document](#document).

Reviewer: Throughout the doc, italics and quotes are used to signify "a term that we established and we'll keep using consistently as something more than its simple English definition". It may help the reader if we use either italics, or bold, or quotes for these, but consistently rather than mixing and matching

Author: Yup, fixed.

## **ID Spaces**

The allocation scheme separates the small-number IDs into two "spaces":

1. **Session space**: IDs are unique within the session scope, but are not guaranteed to be [document](#document) unique without the accompanying context of their session ID.
2. **Op space**: IDs are in their most final form, as unique as possible within the entire system. IDs that have been ordered by the central service ("finalized") are represented in their _final_ form, while any other IDs that have not yet been finalized are left in their _session space_ form.

> Each of these spaces is represented by a discrete type. This allows for the direct encoding of an ID's uniqueness guarantees without contextual ambiguity. An ID can be "normalized" from one space to the other with sychronous call.

Reviewer: ```suggestion
Each of these spaces is represented by a discrete type. This allows for the direct encoding of an ID's uniqueness guarantees without contextual ambiguity. An ID can be "normalized" from one space to the other with a sychronous call.
```

## **Usage Guidelines**

> Allocated IDs can be delivered to application authors for use in place of V4 UUIDs, with the following considerations:

Reviewer: ```suggestion
When delivering Allocated IDs to application authors for use in place of V4 UUIDs, follow these guidelines:
```

- Expose only session space IDs to application authors. This means that application developers don't need to manage multiple ID spaces or worry about converting IDs between session space and op space. This simplifies the data management process and reduces the chances of errors or inconsistencies.

>   - **Note:** Session space IDs are only unique within the originating session (for most applications this will mean within each individual client's app instance, as it is rare to hold handles to more than one allocator). This constrains usage to scopes where IDs are unique; for instance, application developers should never persist these IDs and should instead be directed to convert them to StableIds first.

Reviewer: ```suggestion
  - **Note:** Session space IDs are only unique within the originating session (for most applications this will mean within each individual client's app instance, as it is rare to hold handles to more than one allocator). This constrains usage to scopes where IDs are unique; for instance, application developers should never persist these IDs and should instead be directed to convert them to stable IDs first.
```

> - Use op space IDs for serialized forms, such as sending information over the wire or storing it in a file. Op space IDs are in their _final_ form if possible, which imparts a performance benefit by reducing normalization work when a client receives a network operation or rehydrates an allocator.

Reviewer: ```suggestion
- Use op space IDs for serialized forms, such as sending information over the wire or storing it in a file. Op space IDs are always generated in their _final_ form when possible, which imparts a performance benefit by reducing normalization work when a client receives a network operation or rehydrates an allocator.
```

> - When serializing op space IDs, annotate the entire context (e.g., file or network operation) with the session ID. This is necessary because not all IDs in op space are in their _final_ form. Any non-finalized IDs will still be in their session space form, unique only within that specific session. Annotating the entire context with the session ID ensures that recipients of the serialized data can correctly interpret and process the non-finalized IDs.

Reviewer: Websocket: "Wow, -1, you're so small and lightweight."
Negative Seventeen: "_This isn't even my final form_!"
